src/lib/config.ts

```diff
@@ -0,0 +1,417 @@
+const ini = require("ini");
+
+class InvalidConfigError extends Error {}
+
+// Wraps an `ini` file with helpers to get typed values out
+class Config {
+  keyName: string;
+  private config: { [key: string]: unknown };
+
+  static parse(keyName: string, data: string) {
+    return new Config(keyName, ini.parse(data));
+  }
+
+  constructor(keyName: string, config: { [key: string]: unknown }) {
+    this.keyName = keyName;
+    this.config = config;
+  }
+
+  strOpt(key: string): string | undefined {
+    const v = this.config[key];
+    if (v === undefined || typeof v === "string") {
+      return v;
+    } else {
+      throw new InvalidConfigError(
+        `Expected string for ${this.keyName} ${key}, got ${typeof v}`
+      );
+    }
+  }
+
+  numberOpt(key: string): number | undefined {
+    const v = this.config[key];
+    if (v === undefined || typeof v === "number") {
+      return v;
+    } else if (typeof v === "string") {
+      try {
+        return parseInt(v);
+      } catch (_) {
+        throw new InvalidConfigError(
+          `Invalid number for ${this.keyName} ${key}`
+        );
+      }
+    } else {
+      throw new InvalidConfigError(
+        `Expected number for ${this.keyName} ${key}, got ${typeof v}`
+      );
+    }
+  }
+
+  object(key: string): Config {
+    const v = this.config[key];
+    if (v !== null && typeof v === "object") {
+      return new Config(this.keyName, v as any);
+    } else {
+      throw new InvalidConfigError(
+        `Expected object for ${this.keyName} ${key}, got ${typeof v}`
+      );
+    }
+  }
+
+  str(key: string): string {
+    return this.require(key, this.strOpt(key));
+  }
+
+  number(key: string): number {
+    return this.require(key, this.numberOpt(key));
+  }
+
+  require<T>(key: string, value: T | undefined): T {
+    if (value === undefined) {
+      throw new InvalidConfigError(
+        `Missing value for required ${this.keyName} ${key}`
+      );
+    } else {
+      return value;
+    }
+  }
+
+  entries(pred: (key: string) => boolean): [string, Config][] {
+    return Object.keys(this.config).flatMap((k) =>
+      pred(k) ? [[k, this.object(k)]] : []
+    );
+  }
+}
+
+// Represents `~/.fauna-shell`
+class RootConfig {
+  defaultEndpoint?: string;
+  endpoints: { [key: string]: Endpoint };
+
+  constructor(config: Config) {
+    this.defaultEndpoint = config.strOpt("default");
+    this.endpoints = Object.fromEntries(
+      config
+        .entries((k) => k !== "default")
+        .map(([k, v]) => [k, new Endpoint(v)])
+    );
+
+    if (this.defaultEndpoint === "default") {
+      throw new InvalidConfigError(
+        "Default endpoint cannot be named 'default'"
+      );
+    } else if (
+      this.defaultEndpoint &&
+      this.endpoints[this.defaultEndpoint] === undefined
+    ) {
+      throw new InvalidConfigError(
+        `Default endpoint '${this.defaultEndpoint}' was not found`
+      );
+    }
+  }
+}
+
+// Represents `.fauna-project`
+class LocalConfig {
```

Comment:
could we also name this class ProjectConfig?

Reply:
sure, I can rename to ProjectConfig.

My main hesitation with splitting this up is:
- its not all that large of a file. 400 lines seems perfectly reasonable to me
- typescript makes it difficult to re-export things. so if I defined `ProjectConfig` in `src/lib/config/project.ts`, it would be a bit ugly to write `import { ProjectConfig } from "src/lib/config"` elsewhere.

those are pretty minor though, I'll try reorganizing.